src/main/java/org/opentripplanner/raptor/rangeraptor/TimeBasedRoutingStrategy.java

```diff
@@ -0,0 +1,249 @@
+
+  /**
+   * Get the current boarding previous transit arrival. This is used to look up any guaranteed
+   * transfers.
+   */
+  protected abstract TransitArrival<T> previousTransit(int boardStopIndex);
+
+  @Override
+  public final void board(
+    int stopIndex,
+    int stopPos,
+    int boardSlack,
+    boolean hasConstrainedTransfer,
+    RaptorConstrainedTripScheduleBoardingSearch<T> txSearch
+  ) {
+    // MC Raptor have many, while RR have one boarding
+    forEachBoarding(
+      stopIndex,
+      (int prevArrivalTime) -> {
+        boolean boardedUsingConstrainedTransfer =
+          hasConstrainedTransfer &&
+          boardWithConstrainedTransfer(txSearch, timeTable, stopIndex, prevArrivalTime, boardSlack);
+
+        // Find the best trip and board [no guaranteed transfer exist]
+        if (!boardedUsingConstrainedTransfer) {
+          boardWithRegularTransfer(tripSearch, stopIndex, stopPos, prevArrivalTime, boardSlack);
+        }
+      }
+    );
+  }
+
+  /**
+   * Board trip for each stopArrival (Std have only one "best" arrival, while Mc may have many).
+   */
+  protected abstract void forEachBoarding(int stopIndex, IntConsumer prevStopArrivalTimeConsumer);
+
+  protected final void boardWithRegularTransfer(
+    RaptorTripScheduleSearch<T> tripSearch,
+    int stopIndex,
+    int stopPos,
+    int prevArrivalTime,
+    int boardSlack
+  ) {
+    int earliestBoardTime = earliestBoardTime(prevArrivalTime, boardSlack);
+    // check if we can back up to an earlier trip due to this stop
+    // being reached earlier
+    var result = tripSearch.search(earliestBoardTime, stopPos, onTripIndex());
+    if (result != null) {
+      board(stopIndex, earliestBoardTime, result);
+    } else {
+      boardSameTrip(earliestBoardTime, stopPos, stopIndex);
+    }
+  }
+
+  /**
+   * Board the given trip(event) at the given stop index.
+   *
+   * @param earliestBoardTime used to calculate wait-time (if needed)
+   */
+  protected abstract void board(
+    final int stopIndex,
+    final int earliestBoardTime,
+    RaptorTripScheduleBoardOrAlightEvent<T> boarding
+  );
+
+  /**
+   * The trip search will use this index to search relative to an existing boarding. This make a
+   * subsequent search faster since it must board an earlier trip, and the trip search can start at
+   * the given onTripIndex. if not the current trip is used.
+   * <p>
+   * Return -1 to if the tripIndex is unknown.
+   */
+  protected int onTripIndex() {
+    return -1;
+  }
+
+  /**
+   * This method allow the strategy to replace the existing boarding (if it exists) with a better
+   * option. It is left to the implementation to check that a boarding already exist.
+   *
+   * @param earliestBoardTime - the earliest possible time a boarding can take place
+   * @param stopPos           - the pattern stop position
+   * @param stopIndex         - the global stop index
+   */
+  protected void boardSameTrip(int earliestBoardTime, int stopPos, int stopIndex) {
```

Comment:
This is only used in `MinTravelDurationRoutingStrategy`, should we remove this from `boardWithRegularTransfer` and instead override it in `MinTravelDurationRoutingStrategy`?